feat: Add CSS variables in getImage for experimental.responsiveImages (#13256)
withastro/astro: .changeset/clever-rice-compete.md

---
'astro': patch
---

Adds experimental responsive image support in Markdown

> Previously, the `experimental.responsiveImages` could only provide responsive images when using the `<Image />` and `<Picture />` components.

Reviewer: ```suggestion
Previously, the `experimental.responsiveImages` feature could only provide responsive images when using the `<Image />` and `<Picture />` components.
```

Now, images written with the `![]()` Markdown syntax in Markdown and MDX files will generate responsive images by default when using this experimental feature.

To try this experimental feature, set `experimental.responsiveImages` to true in your `astro.config.mjs` file:

```js
{
   experimental: {
      responsiveImages: true,
   },
}
```

> Learn more about using this feature in the [Experimental responsive images feature reference](https://docs.astro.build/en/reference/experimental-flags/responsive-images/).

Reviewer: ```suggestion
Learn more about using this feature in the [experimental responsive images feature reference](https://docs.astro.build/en/reference/experimental-flags/responsive-images/).
```

---
'astro': patch
---

> Add CSS variables in `getImage` for `experimental.responsiveImages`
> 
> Previously, the responsive images experiment only applied to the `<Image />` and `<Picture />` components. However, images included through Markdown would not have the same styling applied to them, unless using MDX. This change consolidates more styling into `getImage` so it is accessible in more places.

Reviewer: ``````suggestion
Adds experimental responsive image support in Markdown

Previously, the `experimental.responsiveImages` could only provide responsive images when using the `<Image />` and `<Picture />` components.

Now, images written with the `![]()` Markdown syntax in Markdown and MDX files will generate responsive images by default when using this experimental feature.

To try this experimental feature, set `experimental.responsiveImages` to true in your `astro.config.mjs` file:

```js
{
   experimental: {
      responsiveImages: true,
   },
}
```

Learn more about using this feature in the [Experimental responsive images feature reference](https://docs.astro.build/en/reference/experimental-flags/responsive-images/).

For a complete overview, and to give feedback on this experimental API, see the [Responsive Images RFC](https://github.com/withastro/roadmap/blob/responsive-images/proposals/0053-responsive-images.md).
``````

So, for an experimental feature, it's probably good to remind people how they can enable it, if they haven't already done so! How about something like the above?

Also, I don't see any docs PR for this yet. What I'd need to confirm to be able to update the docs page is:

- Since you can't pass props in Markdown syntax the way you can with the `<Image />` and `<Picture />` components, is it correct to say that all of these images will just use whatever is configured in the `image` config object? (e.g. `image.experimentalLayout`, `image.experimentalObjectFit`, etc.?) And therefore, something like `priority` which has no config setting just defaults to its own default and cannot be changed?

I'll note that the docs page right now is **heavily geared towards the two components** (e.g. "When this flag is enabled, you can pass a layout prop to any `<Image />` or `<Picture />` component to create a responsive image." So, it will take a bit of a surgical rewrite of the page to spot things like this throughout.

**I am happy to tackle this docs work** if you confirm the above! It would probably be easier than me asking anyone else to try, then just changing it myself anyway. 😅

Author: > Since you can't pass props in Markdown syntax the way you can with the `<Image />` and `<Picture />` components, is it correct to say that all of these images will just use whatever is configured in the `image` config object? (e.g. `image.experimentalLayout`, `image.experimentalObjectFit`, etc.?) And therefore, something like `priority` which has no config setting just defaults to its own default and cannot be changed?

This is true. Technically, you can write a rehype plugin to hook into Astro internals [like I do in my site](https://github.com/p0lyw0lf/website/blob/40e7a89463d5af55f2ea99d7a21f64a702e957e7/src/plugins/rehypeRemoteImages.ts#L41-L48) to change these values, but this is probably not worth mentioning in the docs.

Thank you for all your hard work making Astro's docs shine btw! A big reason why I'm enamored with this project.

Reviewer: Great, thanks! I *think* that's all I need to know!

I'll loop you in for a review to catch anything I might have got wrong! Note: these astro PRs get merged on Wednesdays, but docs doesn't need to be finalized until Thursday because our merges go live immediately, not into a release. So we have an extra day and don't be surprised if I take it! It's not a problem with getting the feature merged!
